domain/src/main/java/org/oppia/android/domain/topic/TopicListController.kt

```diff
@@ -143,15 +203,247 @@ class TopicListController @Inject constructor(
+
+          val startedChapterProgressList =
+            storyProgress.chapterProgressMap.values
+              .filter { chapterProgress ->
+                chapterProgress.chapterPlayState ==
+                  ChapterPlayState.STARTED_NOT_COMPLETED
+              }
+              .sortedByDescending { chapterProgress -> chapterProgress.lastPlayedTimestamp }
+
+          val recentlyPlayerChapterProgress: ChapterProgress? =
+            startedChapterProgressList.firstOrNull()
+          if (recentlyPlayerChapterProgress != null) {
+            val recentlyPlayerChapterSummary: ChapterSummary? =
+              story.chapterList.find { chapterSummary ->
+                recentlyPlayerChapterProgress.explorationId == chapterSummary.explorationId
+              }
+            if (recentlyPlayerChapterSummary != null) {
+              val numberOfDaysPassed =
+                (Date().time - recentlyPlayerChapterProgress.lastPlayedTimestamp) / ONE_DAY_IN_MS
+              val promotedStory = createPromotedStory(
+                storyId,
+                topic,
+                completedChapterProgressList.size,
+                story.chapterCount,
+                recentlyPlayerChapterSummary.name,
+                recentlyPlayerChapterSummary.explorationId
+              )
+              if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
+                ongoingStoryListBuilder.addRecentStory(promotedStory)
+              } else {
+                ongoingStoryListBuilder.addOlderStory(promotedStory)
+              }
+            }
+          } else if (lastCompletedChapterProgress != null &&
+            lastCompletedChapterProgress.explorationId != story.chapterList.last().explorationId
+          ) {
+            val lastChapterSummary: ChapterSummary? = story.chapterList.find { chapterSummary ->
+              lastCompletedChapterProgress.explorationId == chapterSummary.explorationId
+            }
+            val nextChapterIndex = story.chapterList.indexOf(lastChapterSummary) + 1
+            val nextChapterSummary: ChapterSummary? = story.chapterList[nextChapterIndex]
+            if (nextChapterSummary != null) {
+              val numberOfDaysPassed =
+                (Date().time - lastCompletedChapterProgress.lastPlayedTimestamp) / ONE_DAY_IN_MS
+              val promotedStory = createPromotedStory(
+                storyId,
+                topic,
+                completedChapterProgressList.size,
+                story.chapterCount,
+                nextChapterSummary.name,
+                nextChapterSummary.explorationId
+              )
+              if (numberOfDaysPassed < ONE_WEEK_IN_DAYS) {
+                ongoingStoryListBuilder.addRecentStory(promotedStory)
+              } else {
+                ongoingStoryListBuilder.addOlderStory(promotedStory)
+              }
+            }
+          }
+        }
+      }
+    }
+
+    return ongoingStoryListBuilder.build()
+  }
+
+  private fun createRecommendedActivityList(
```

Comment:
Ditto for this (please split it up--it's getting hard to follow the code flow here).

Reply:
Done.